examples/tool_chat_template_hermes.jinja

```diff
@@ -0,0 +1,120 @@
+{%- macro json_to_python_type(json_spec) %}
+{%- set basic_type_map = {
+    "string": "str",
+    "number": "float",
+    "integer": "int",
+    "boolean": "bool"
+} %}
+
+{%- if basic_type_map[json_spec.type] is defined %}
+    {{- basic_type_map[json_spec.type] }}
+{%- elif json_spec.type == "array" %}
+    {{- "list[" +  json_to_python_type(json_spec|items) + "]"}}
+{%- elif json_spec.type == "object" %}
+    {%- if json_spec.additionalProperties is defined %}
+        {{- "dict[str, " + json_to_python_type(json_spec.additionalProperties) + ']'}}
+    {%- else %}
+        {{- "dict" }}
+    {%- endif %}
+{%- elif json_spec.type is iterable %}
+    {{- "Union[" }}
+    {%- for t in json_spec.type %}
+      {{- json_to_python_type({"type": t}) }}
+      {%- if not loop.last %}
+        {{- "," }} 
+    {%- endif %}
+    {%- endfor %}
+    {{- "]" }}
+{%- else %}
+    {{- "Any" }}
+{%- endif %}
+{%- endmacro %}
+
+
+{{- bos_token }}
+{{- "You are a function calling AI model. You are provided with function signatures within <tools></tools> XML tags. You may call one or more functions to assist with the user query. Don't make assumptions about what values to plug into functions. Here are the available tools: <tools> " }}
```

Comment:
What about wrapping the tool block (from line 35 to 84) in a

{%- if tools is iterable and len(tools) > 0 %}

{%- endif %}

to behave better when to tool list is empty or not defined ?

Reply:
This template is the official template hermes models provided by Nous Research, with one modification that they reviewed & thought was ok. Happy to do this if they don't have an issue with it. @interstellarninja ?

Reply:
yes it makes sense to skip the step between line 35 and 84 if tools were not passed when making chat completion request, we may still need to default to another system prompt? 

how do Mistral or CommandR models handle non-tool call system prompt?

but if tools were passed but not well defined, do we just want to return a warning (if not an error) and still use the template with empty list between `<tools> [] </tools>` tags?

@K-Mistele do you know how huggingface handles this case?
https://github.com/huggingface/transformers/blob/main/src/transformers/tokenization_utils_base.py#L1687

Reply:
I don't know how it handles that case since it's definitely an edge case. I have never seen that. It would probably take a lot of prompting to get the model to do that. Maybe we just return an empty tool array, as well as any text content that the model generated?

Reply:
I seem have found an exact problem with these lines while using Autogen groupchats. 

The autogen groupchat manager specifically triggers the following error because the groupchat manager is passing "None" to the template for the tools variable by design. 

`openai.BadRequestError: Error code: 400 - {'object': 'error', 'message': "'NoneType' object is not iterable", 'type': 'BadRequestError', 'param': None, 'code': 400}`


Autogen specifically leaves out function calling from the responsibilities of the Groupchat Manager, as shown by the lack of tool choice in the llm_config for it in the example below, as it is tasked with picking the next speaker in the group only.

If I try to wrap the template section like the following to handle a case where tools is "None" like the following:

```
{{- bos_token }}
{{- "You are a function calling AI model. You are provided with function signatures within <tools></tools> XML tags. You may call one or more functions to assist with the user query. Don't make assumptions about what values to plug into functions. Here are the available tools: <tools>" }}
{%- if tools %}
    {%- for tool in tools %}
        {# Existing tool processing logic #}
    {%- endfor %}
{%- else %}
    {{- " No tools are currently available. " }}
{%- endif %}
{{- "</tools>" }}
```

I get this following error:

```
  File "/home/wybartel/.local/lib/python3.10/site-packages/uvicorn/protocols/http/httptools_impl.py", line 426, in run_asgi
    result = await app(  # type: ignore[func-returns-value]
  File "/home/wybartel/.local/lib/python3.10/site-packages/uvicorn/middleware/proxy_headers.py", line 84, in __call__
    return await self.app(scope, receive, send)
  File "/home/wybartel/miniconda3/envs/vllm/lib/python3.10/site-packages/fastapi/applications.py", line 1054, in __call__
    await super().__call__(scope, receive, send)
  File "/home/wybartel/miniconda3/envs/vllm/lib/python3.10/site-packages/starlette/applications.py", line 123, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/home/wybartel/miniconda3/envs/vllm/lib/python3.10/site-packages/starlette/middleware/errors.py", line 186, in __call__
    raise exc
  File "/home/wybartel/miniconda3/envs/vllm/lib/python3.10/site-packages/starlette/middleware/errors.py", line 164, in __call__
    await self.app(scope, receive, _send)
  File "/home/wybartel/miniconda3/envs/vllm/lib/python3.10/site-packages/starlette/middleware/cors.py", line 85, in __call__
    await self.app(scope, receive, send)
  File "/home/wybartel/miniconda3/envs/vllm/lib/python3.10/site-packages/starlette/middleware/exceptions.py", line 65, in __call__
    await wrap_app_handling_exceptions(self.app, conn)(scope, receive, send)
  File "/home/wybartel/miniconda3/envs/vllm/lib/python3.10/site-packages/starlette/_exception_handler.py", line 64, in wrapped_app
    raise exc
  File "/home/wybartel/miniconda3/envs/vllm/lib/python3.10/site-packages/starlette/_exception_handler.py", line 53, in wrapped_app
    await app(scope, receive, sender)
  File "/home/wybartel/miniconda3/envs/vllm/lib/python3.10/site-packages/starlette/routing.py", line 756, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/home/wybartel/miniconda3/envs/vllm/lib/python3.10/site-packages/starlette/routing.py", line 776, in app
    await route.handle(scope, receive, send)
  File "/home/wybartel/miniconda3/envs/vllm/lib/python3.10/site-packages/starlette/routing.py", line 297, in handle
    await self.app(scope, receive, send)
  File "/home/wybartel/miniconda3/envs/vllm/lib/python3.10/site-packages/starlette/routing.py", line 77, in app
    await wrap_app_handling_exceptions(app, request)(scope, receive, send)
  File "/home/wybartel/miniconda3/envs/vllm/lib/python3.10/site-packages/starlette/_exception_handler.py", line 64, in wrapped_app
    raise exc
  File "/home/wybartel/miniconda3/envs/vllm/lib/python3.10/site-packages/starlette/_exception_handler.py", line 53, in wrapped_app
    await app(scope, receive, sender)
  File "/home/wybartel/miniconda3/envs/vllm/lib/python3.10/site-packages/starlette/routing.py", line 72, in app
    response = await func(request)
  File "/home/wybartel/miniconda3/envs/vllm/lib/python3.10/site-packages/fastapi/routing.py", line 278, in app
    raw_response = await run_endpoint_function(
  File "/home/wybartel/miniconda3/envs/vllm/lib/python3.10/site-packages/fastapi/routing.py", line 191, in run_endpoint_function
    return await dependant.call(**values)
  File "/home/wybartel/dev/vllm/vllm/entrypoints/openai/api_server.py", line 132, in create_chat_completion
    generator = await openai_serving_chat.create_chat_completion(
  File "/home/wybartel/dev/vllm/vllm/entrypoints/openai/serving_chat.py", line 375, in create_chat_completion
    generator = await self.chat_completion_full_generator(
  File "/home/wybartel/dev/vllm/vllm/entrypoints/openai/serving_chat.py", line 698, in chat_completion_full_generator
    message=message,
*UnboundLocalError: local variable 'message' referenced before assignment*
```



Example code from the Autogen examples is provided here and converted from the openai calls to pointing at vllm, though an env file that looks something like this is required to run:

`pip install pyautogen`

```
#autogen groupchat example.py
import os
import json
import time
from typing_extensions import Annotated
import autogen
import requests
from autogen.cache import Cache

# Retrieve the env variables
model = "NousResearch/Hermes-2-Pro-Llama-3-70B"
api_endpoint = "http://192.168.4.10:8000"

openai_api_base = api_endpoint + '/v1'

# Define the configuration list
config_list = [{
    "model": model,
    "base_url": openai_api_base,
    "api_key": "NULL",  # Placeholder
    "extra_body": {"stop_token_ids": [128001, 128009]},
}]

llm_config = {
    "config_list": config_list,
    "timeout": 30000,
    "tool_choice": "auto",
}

llm_config_manager = {
    "config_list": config_list,
    "timeout": 30000,
}

# Create agents
coder = autogen.AssistantAgent(
    name="chatbot",
    system_message="For coding tasks, only use the functions you have been provided with. You have a stopwatch and a timer, these tools can and should be used in parallel. Reply TERMINATE when the task is done.",
    llm_config=llm_config,
)

user_proxy = autogen.UserProxyAgent(
    name="user_proxy",
    system_message="A proxy for the user for executing code.",
    is_termination_msg=lambda x: x.get("content", "") and x.get("content", "").rstrip().endswith("TERMINATE"),
    human_input_mode="NEVER",
    max_consecutive_auto_reply=10,
    code_execution_config={"work_dir": "coding"},
)

# Define functions
@user_proxy.register_for_execution()
@coder.register_for_llm(description="create a timer for N seconds")
def timer(num_seconds: Annotated[str, "Number of seconds in the timer."]) -> str:
    for i in range(int(num_seconds)):
        time.sleep(1)
    return "Timer is done!"

def stopwatch(num_seconds: Annotated[str, "Number of seconds in the stopwatch."]) -> str:
    for i in range(int(num_seconds)):
        time.sleep(1)
    return "Stopwatch is done!"

autogen.agentchat.register_function(
    stopwatch,
    caller=coder,
    executor=user_proxy,
    description="create a stopwatch for N seconds",
)

# Group chat setup
markdownagent = autogen.AssistantAgent(
    name="Markdown_agent",
    system_message="Respond in markdown only",
    llm_config=llm_config,
)

@user_proxy.register_for_execution()
@markdownagent.register_for_llm()
@coder.register_for_llm(description="terminate the group chat")
def terminate_group_chat(message: Annotated[str, "Message to be sent to the group chat."]) -> str:
    return f"[GROUPCHAT_TERMINATE] {message}"

groupchat = autogen.GroupChat(agents=[user_proxy, coder, markdownagent], messages=[], max_round=12)

manager = autogen.GroupChatManager(
    groupchat=groupchat,
    llm_config=llm_config_manager,
    is_termination_msg=lambda x: "GROUPCHAT_TERMINATE" in x.get("content", ""),
)

# Group chat initialization
def group_chat_main():
    message = """
    1) Create a timer and a stopwatch for 5 seconds each in parallel.
    2) Pretty print the result as md.
    3) when 1 and 2 are done, terminate the group chat
    """
    user_proxy.initiate_chat(
        manager,
        message=message,
    )

# Run the main function
if __name__ == "__main__":
    group_chat_main()
```

Reply:
> passing "None" to the template for the tools variable by design
if tools are not passed we can skip using the tool-use system prompt

but if tools are passed incorrectly, we should just throw an error!

Reply:
@wybartel exactly the same problem as you. The error you see is related to my other comment https://github.com/vllm-project/vllm/pull/5649#discussion_r1684689497.

 I think it is a bad practice to let a variable (here `message`) to be potentially undefined after a bunch of conditions. If the behavior is unwanted a proper exception should be raised. Otherwise the variable should be initialize to some default (see my coment).

Reply:
Handled initializing message properly in this case in 95605912. I will add the conditional intot he prompt template as well.

Reply:
b2ceb71f